libraries/AP_Math/AP_Math.cpp

```diff
@@ -49,3 +66,108 @@ float linear_interpolate(float low_output, float high_output,
     return low_output + p * (high_output - low_output);
 }
 
+template <class T>
+float wrap_180(const T angle, float unit_mod)
+{
+    auto res = wrap_360(angle, unit_mod);
+    if (res > 180.f * unit_mod) {
+        res -= 360.f * unit_mod;
+    }
+    return res;
+}
+
+template float wrap_180<int>(const int angle, float unit_mod);
+template float wrap_180<short>(const short angle, float unit_mod);
+template float wrap_180<float>(const float angle, float unit_mod);
+template float wrap_180<double>(const double angle, float unit_mod);
+
+template <class T>
+auto wrap_180_cd(const T angle) -> decltype(wrap_180(angle, 100.f))
+{
+    return wrap_180(angle, 100.f);
+}
+
+template auto wrap_180_cd<float>(const float angle) -> decltype(wrap_180(angle, 100.f));
+template auto wrap_180_cd<int>(const int angle) -> decltype(wrap_180(angle, 100.f));
+template auto wrap_180_cd<short>(const short angle) -> decltype(wrap_180(angle, 100.f));
+template auto wrap_180_cd<double>(const double angle) -> decltype(wrap_360(angle, 100.f));
+
+template <class T>
+float wrap_360(const T angle, float unit_mod)
+{
+    const float ang_360 = 360.f * unit_mod;
+    float res = fmodf(static_cast<float>(angle), ang_360);
+    if (res < 0) {
+        res += ang_360;
+    }
+    return res;
+}
+
+template float wrap_360<int>(const int angle, float unit_mod);
+template float wrap_360<short>(const short angle, float unit_mod);
+template float wrap_360<float>(const float angle, float unit_mod);
+template float wrap_360<double>(const double angle, float unit_mod);
+
+template <class T>
+auto wrap_360_cd(const T angle) -> decltype(wrap_360(angle, 100.f))
+{
+    return wrap_360(angle, 100.f);
+}
+
+template auto wrap_360_cd<float>(const float angle) -> decltype(wrap_360(angle, 100.f));
+template auto wrap_360_cd<int>(const int angle) -> decltype(wrap_360(angle, 100.f));
+template auto wrap_360_cd<short>(const short angle) -> decltype(wrap_360(angle, 100.f));
+template auto wrap_360_cd<double>(const double angle) -> decltype(wrap_360(angle, 100.f));
+
+template <class T>
+float wrap_PI(const T radian)
+{
+    auto res = wrap_2PI(radian);
+    if (res > M_PI) {
+        res -= M_2PI;
+    }
+    return res;
+}
+
+template float wrap_PI<int>(const int radian);
+template float wrap_PI<short>(const short radian);
+template float wrap_PI<float>(const float radian);
+template float wrap_PI<double>(const double radian);
+
+template <class T>
+float wrap_2PI(const T radian)
+{
+    float res = fmodf(static_cast<float>(radian), M_2PI);
+    if (res < 0) {
+        res += M_2PI;
+    }
+    return res;
+}
+
+template float wrap_2PI<int>(const int radian);
+template float wrap_2PI<short>(const short radian);
+template float wrap_2PI<float>(const float radian);
+template float wrap_2PI<double>(const double radian);
+
+template <class T>
+T constrain_value(const T amt, const T low, const T high)
+{
+    if (isnan(low) || isnan(high)) {
```

Comment:
In the current _constrain_float_ we check if _amt_ was NaN (but not _low_ and _high_) and here we check _low_ and _high_ (but not _amt_). Is there any reason for this change? The current version even has comment saying the check for NaN is to prevent propagation of floating point errors - it is an old comment (2013), not sure if still applies.